src/object/basic/co_integer48.c

```diff
@@ -0,0 +1,121 @@
+   limitations under the License.
+******************************************************************************/
+
+/******************************************************************************
+* INCLUDES
+******************************************************************************/
+
+#include "co_core.h"
+
+/******************************************************************************
+* PRIVATE DEFINES
+******************************************************************************/
+
+#define COT_ENTRY_SIZE    (uint32_t)6
+
+/******************************************************************************
+* PRIVATE FUNCTIONS
+******************************************************************************/
+
+static uint32_t COTInt48Size (struct CO_OBJ_T *obj, struct CO_NODE_T *node, uint32_t width);
+static CO_ERR   COTInt48Read (struct CO_OBJ_T *obj, struct CO_NODE_T *node, void *buffer, uint32_t size);
+static CO_ERR   COTInt48Write(struct CO_OBJ_T *obj, struct CO_NODE_T *node, void *buffer, uint32_t size);
+
+/******************************************************************************
+* PUBLIC GLOBALS
+******************************************************************************/
+
+const CO_OBJ_TYPE COTInt48 = { COTInt48Size, 0, COTInt48Read, COTInt48Write, 0 };
+
+/******************************************************************************
+* FUNCTIONS
+******************************************************************************/
+
+static uint32_t COTInt48Size(struct CO_OBJ_T *obj, struct CO_NODE_T *node, uint32_t width)
+{
+    uint32_t result = (uint32_t)0;
+
+    CO_UNUSED(node);
+    CO_UNUSED(width);
+
+    if (CO_IS_DIRECT(obj->Key) != 0) {
+        result = (uint32_t)COT_ENTRY_SIZE;
+    } else {
+        /* check for valid reference */
+        if ((obj->Data) != (CO_DATA)0) {
+            result = (uint32_t)COT_ENTRY_SIZE;
+        }
+    }
+    return (result);
+}
+
+static CO_ERR COTInt48Read(struct CO_OBJ_T *obj, struct CO_NODE_T *node, void *buffer, uint32_t size)
+{
+    CO_ERR   result = CO_ERR_NONE;
+    uint64_t value;
+
+    CO_UNUSED(node);
+    ASSERT_PTR_ERR(obj, CO_ERR_BAD_ARG);
+    ASSERT_PTR_ERR(buffer, CO_ERR_BAD_ARG);
+
+    if (CO_IS_DIRECT(obj->Key) != 0) {
+        value = (uint64_t)(obj->Data);
+    } else {
+        value = *((uint64_t *)(obj->Data));
+    }
+    if (CO_IS_NODEID(obj->Key) != 0) {
+        value += node->NodeId;
+    }
+    if (size == COT_ENTRY_SIZE) {
+        *((uint64_t *)buffer) = value;
+    } else {
+        result = CO_ERR_BAD_ARG;
+    }
+    return (result);
+}
+
+static CO_ERR COTInt48Write(struct CO_OBJ_T *obj, struct CO_NODE_T *node, void *buffer, uint32_t size)
+{
+    CO_ERR   result = CO_ERR_NONE;
+    uint64_t value;
+    uint64_t oldValue;
+
+    CO_UNUSED(node);
+    ASSERT_PTR_ERR(obj, CO_ERR_BAD_ARG);
+    ASSERT_PTR_ERR(buffer, CO_ERR_BAD_ARG);
+
+    value = *((uint64_t *)buffer);

```

Comment:
To get the signed and unsigned version managed correctly, we should add:

```c
value = *((uint64_t *)buffer);
if ((value & (uint64_t)0x0000800000000000) != 0) {
    /* fill sign-bits for negative 48bit values */
    value |= (uint64_t)0xFFFF000000000000;
} else {
    /* limit positive value to 48bit */
    value &= (uint64_t)0x0000FFFFFFFFFFFF;
}
if (size == COT_ENTRY_SIZE) {
  :
```

This handles the unused bits according to positive or negative values.

Reply:
You're right, I've only dealt with UNSIGNED48 but I test this then with INTEGER48.
I have a CANopen analyzer by emotas, so I can fill a EDS file and add an INTEGER48 signed into the dictionary and do a SDO transfer and see if it works correctly with this code. If yes I add it to the patch and push --force.

Reply:
Ok.